### PR TITLE
sema: check input patch size mismatch earlier.

### LINKED ISSUE #3739: [SPIR-V] No error when Hull and patchconstantfunc input point counts don't match
### Title
[SPIR-V] No error when Hull and patchconstantfunc input point counts don't match (errors with DXIL backend)

### Functional impact
The InputPatch point count argument is ignored for the patch constant function. This will result in undetected/silent errors, when a shader writer might believe they modified the input point count but actually don't.

### Minimal repro steps

1. Compile the following shader with `dxc.exe -T hs_6_0 -E Hull -spirv repro_shader.txt`:
[repro_shader.txt](https://github.com/microsoft/DirectXShaderCompiler/files/6415304/repro_shader.txt)

### Expected result
Compilation should fail with error as produced by the DXIL backend:
```
repro_shader.txt:7:14: error: Patch constant function's input patch input should have 3 elements, but has 16.
HullPatchOut HullConst (InputPatch<ControlPoint,16> v) { return (HullPatchOut)0; }
             ^
```
### Actual result
Compilation completes with the point count in the patch constant function being ignored (cannot find the patch function's input point count in the generated SPIR-V)

### Further technical details
Used DXC release v1.6.2104 (e09a454)

### PR DESCRIPTION
**Pull request by Keenuts**
The check was done in the DXIL backend, and not in the SPIR-V backend. This causes the issue to be ignored on the SPIR-V side. Moved the check earlier to SEMA so a single check works for both backends.

I had to split the existing HLSL check as it doesn't support having 2 run lines (AFAIK).

Fixes #3739

### CODE REVIEW

File: tools/clang/test/HLSLFileCheck/samples/SimpleHs12.hlsl [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+// hull per-control point shader
+[domain("tri")]
+[partitioning("fractional_odd")]
+[outputtopology("triangle_cw")]
+[patchconstantfunc("HSPerPatchFunc")]
+[outputcontrolpoints(3)]
+HSPerVertexData main(const uint id : SV_OutputControlPointID,
+                      const InputPatch< PSSceneIn, 12 > points )
```
Comment (bob80905) [inline tools/clang/test/HLSLFileCheck/samples/SimpleHs12.hlsl:67]: Would it not be possible to change the input patch number here, to catch both diagnostics at the same time with one run line?

Comment (Keenuts) [reply tools/clang/test/HLSLFileCheck/samples/SimpleHs12.hlsl:67]: When I tried it didn't work, as the error checked in the other file makes the compile process abort earlier, and the 2 other messages don't show.

Comment (sudonatalie) [reply tools/clang/test/HLSLFileCheck/samples/SimpleHs12.hlsl:67]: It seems a bit odd that these are caught at different points now so I'd probably want to move all 3 checks to Sema, or at least track the remaining 2 in a follow-up bug, but I'll leave it to MS folks to decide.

Comment (Keenuts) [reply tools/clang/test/HLSLFileCheck/samples/SimpleHs12.hlsl:67]: Created 2 issues to track those. One is handled by the SPIR-V backend, but weirdly, the other also causes us to accept bad code.

**Review by bob80905 — APPROVED** [review]

**Pull request merged**